libafl_nyx/src/helper.rs

```diff
@@ -0,0 +1,124 @@
+/// [`NyxHelper`] is used to wrap `NyxProcess`
+use std::{
+    fmt::{self, Debug},
+    path::Path,
+};
+
+use libafl::Error;
+use libnyx::{NyxProcess, NyxReturnValue};
+
+pub struct NyxHelper {
+    pub nyx_process: NyxProcess,
+    /// real size of trace_bits
+    pub real_map_size: usize,
+    // real size of the trace_bits
+    pub map_size: usize,
+    /// shared memory with instruction bitmaps
+    pub trace_bits: *mut u8,
+}
+
+const MAX_FILE: u32 = 1024 * 1024;
+#[derive(Clone, Copy)]
+pub enum NyxProcessType {
+    /// stand alone mode
+    ALONE,
+    /// parallel mode's parent, used to create snapshot
+    PARENT,
+    /// parallel mode's child, consume snapshot and execute
+    CHILD,
+}
+impl NyxHelper {
+    /// create `NyxProcess` and do basic settings
+    /// It will convert instance to parent or child using `parent_cpu_id` when set`parallel_mode`
+    pub fn new(
+        target_dir: &Path,
+        cpu_id: u32,
+        snap_mode: bool,
+        parallel_mode: bool,
+        parent_cpu_id: Option<u32>,
+    ) -> Result<Self, Error> {
+        let sharedir = target_dir
+            .to_str()
+            .expect("unable to convert target_dir to str");
+        let work_dir = target_dir.join("workdir");
+        let work_dir = work_dir.to_str().expect("unable to convert workdir to str");
+        let nyx_type = if parallel_mode {
+            let parent_cpu_id = match parent_cpu_id {
```

Comment:
Why not always try to become the `parent`, and fall back to `Child` on error? Would make the API a lot cleaner/easier to use

Reply:
I have tried, but the error is thrown when spawning the process instead of `NyxProcess::new_parent`, so it will panic instead of returning an error.
By the way, the cpu_id itself can also serve as `work_id` to distinguish each other. It helps to eliminate the hack of sharing a count for spawned instances with `NyxHelper`.

Reply:
Should we ask the nyx authors to return `Result` instead of panic? Could be a cleaner API. cc @eqv

Reply:
For now you could also consider catching the panic: https://doc.rust-lang.org/std/panic/fn.catch_unwind.html

Reply:
I have opened https://github.com/nyx-fuzz/libnyx/issues/15 for this